adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/impl/MessageMapping.java

```diff
@@ -0,0 +1,115 @@
+     * @return Mapped message
+     */
+    public Future<MappedMessage> mapMessage(final MqttContext ctx, final ResourceIdentifier targetAddress,
+                                     final MqttPublishMessage message, final JsonObject registrationInfo, final MqttProtocolAdapterProperties config) {
+        final Promise<MappedMessage> result = Promise.promise();
+        if (!registrationInfo.containsKey(RegistrationConstants.FIELD_MAPPER)) {
+            result.complete(new MappedMessage(ctx, targetAddress, message));
+            return result.future();
+        }
+        final String mapper = registrationInfo.getString(RegistrationConstants.FIELD_MAPPER);
+        if (mapper.isBlank()) {
+            result.complete(new MappedMessage(ctx, targetAddress, message));
+            return result.future();
+        }
+
+        final MapperEndpoint mapperEndpoint = config.getMapperEndpoint(mapper);
+        if (mapperEndpoint == null) {
+            result.complete(new MappedMessage(ctx, targetAddress, message));
+            return result.future();
+        }
+
+        final VertxHttpHeaders headers = new VertxHttpHeaders();
+        registrationInfo.iterator().forEachRemaining(stringObjectEntry -> headers.add(stringObjectEntry.getKey(),
+                stringObjectEntry.getValue()));
+
+        return mapMessageRequest(ctx, targetAddress, message, mapperEndpoint, headers);
+    }
+
+    Future<MappedMessage> mapMessageRequest(final MqttContext ctx, final ResourceIdentifier targetAddress,
+                                            final MqttPublishMessage message, final MapperEndpoint mapperEndpoint, final VertxHttpHeaders headers) {
+        final Promise<MappedMessage> result = Promise.promise();
+        webClient.post(mapperEndpoint.getPort(), mapperEndpoint.getHost(), mapperEndpoint.getUri())
+                .putHeaders(headers)
+                .ssl(mapperEndpoint.ssl)
+                .sendBuffer(message.payload(), httpResponseAsyncResult -> {
+                    if (httpResponseAsyncResult.succeeded()) {
+                        final HttpResponse<Buffer> httpResponse = httpResponseAsyncResult.result();
+                        if (httpResponse.statusCode() == 200) {
+                            final MqttPublishMessageImpl mqttPublishMessage = new MqttPublishMessageImpl(ctx.message().messageId(), ctx.message().qosLevel(),
+                                    ctx.message().isDup(), ctx.message().isRetain(), ctx.message().topicName(),
+                                    httpResponse.bodyAsBuffer().getByteBuf());
+                            final MultiMap responseHeaders = httpResponse.headers();
+                            String deviceId = targetAddress.getResourceId();
+                            if (headers.contains(MessageHelper.APP_PROPERTY_DEVICE_ID)) {
+                                deviceId = headers.get(MessageHelper.APP_PROPERTY_DEVICE_ID);
+                            }
```

Comment:
If I am not mistaken this should try to retrieve the (mapped) device ID from the response headers, shouldn't it?